### ISSUE TITLE
PYI034 false positive if Self is a quoted annotation

### ISSUE THREAD

**Issue by BenGale93** [astral-sh/ruff#12928]
List of keywords searched:
* PYI034

Code example:

```python
import typing as t

if t.TYPE_CHECKING:
    from typing_extensions import Self


class Foo:
    def __init__(self, x: int) -> None:
        self.x = x

    def __iadd__(self, other: "Self") -> "Self":
        """Adds two Foo objects together."""
        self.x += other.x
        return self
```

Command: `ruff check file.py --isolated --select=PYI034`

Ruff Version: 0.5.7

I'd be happy to try and fix this as I have some Rust experience. Please let me know if I can be of help.

**Comment by AlexWaygood** [comment]
> I'd be happy to try and fix this as I have some Rust experience. Please let me know if I can be of help.

Please feel free! We have logic elsewhere for handling quoted annotations, so I'd definitely go and check how other rules do this kind of thing first

**Comment by BenGale93** [comment]
I think this might be more of a fundamental issue.

If I extend the example:

```python
import typing as t

if t.TYPE_CHECKING:
    from typing import Any

    from typing_extensions import Self


class Foo:
    def __init__(self, x: int) -> None:
        self.x = x

    def __iadd__(self, other: "Self") -> "Self":
        self.x += other.x
        return self

    def __eq__(self, value: "Any") -> bool:
        return True
```

PYI031 should pick up the `Any` in the `__eq__` but it doesn't. Appreciate the example is a little contrived.

I've been combing through the code on how to resolve it but I'm struggling. Any suggestions?

**Comment by AlexWaygood** [comment]
Hmm, yeah, this was a bit more complicated than I realised. Unfortunately in order to figure out how to fix it, I've basically had to go ahead and implement the fix locally, so I might just go ahead and file a PR myself if that's okay :/ Sorry -- I thought this was a good contributor issue but I think I was mistaken.

**Comment by AlexWaygood** [comment]
I've filed #12951

**Comment by BenGale93** [comment]
Thanks @AlexWaygood! No worries about taking over, I'm certain I would have never come up with that fix.
